modules/host/contractcreation.go

```diff
@@ -129,25 +133,60 @@ func (h *Host) considerContract(t consensus.Transaction) (updatedTransaction con
 // reason, all of the locking in this function is done manually. Edit with
 // caution, review with caution.
 //
-// TODO: Split some of this logic into helper functions, which in particular
-// will make mutex management easier.
+// TODO: Order of events: (not currently followed)
+//			1. Contract is sent over without signatures, and doesn't need an
+//			acurate merkle hash of the file. It's just for the host and client
+//			to confirm that the terms of the contract are correct and make
+//			sens.
+//
+//			2. Host sends a confirmation, along with an indication of how much
+//			the client should contribute for payment, and how much is available
+//			for burn.
+//
+//			3. File is sent over and loaded onto the host's disk. Eventually,
+//			micropayments will be worked into this situation to pay the host
+//			for bandwidth and whatever storage time.
+//
+//			4. Client sends over the final client version of the contract
+//			containing the appropriate price and burn, with signatures and the
+//			correct merkle root hash.
+//
+//			5. Host compares the contract with the earlier negotiations and
+//			confirms that everything is still the same. Host then verifies that
+//			the file which was uploaded has a matching hash to the file in the
+//			contract. Host will add burn dollars and submit the contract to the
+//			blockchain.
+//
+//			6. If the client double-spends, the host will remove the file. If
+//			the host holds the transaction hostage and does not submit it to
+//			the blockchain, it will become void.
+//
+// TODO: This functions error handling isn't safe, reveals too much info to the
+// other party.
 func (h *Host) NegotiateContract(conn net.Conn) (err error) {
 	// Read the transaction from the connection.
-	var t consensus.Transaction
-	if err = encoding.ReadObject(conn, &t, maxContractLen); err != nil {
+	var txn consensus.Transaction
+	err = encoding.ReadObject(conn, &txn, maxContractLen)
+	if err != nil {
+		return
+	}
+	var startBlock consensus.BlockHeight
+	err = encoding.ReadObject(conn, &startBlock, maxContractLen) // what should be the maxlen here?
+	if err != nil {
 		return
 	}
+	contract := txn.FileContracts[0]
 
 	// Check that the contained FileContract fits host criteria for taking
 	// files, replying with the error if there's a problem.
 	h.mu.Lock()
-	t, err = h.considerContract(t)
+	txn, err = h.considerContract(txn, startBlock)
```

Comment:
no :(

At some point, `considerContract` is actually going to take a set of contracts and consider them as 1. Contracts got changed so that there's only 1 output. But hosts would prefer to get paid every N blocks right, so you'll have a transaction with like 12 contracts, each for longer and longer durations (one finalizes at block 100, one at 200, 300, etc.) but all paying the same price. But since it's the same hash, the host knows that it's fair.

It's going to get more complicated, need to think of the simplest way to achieve all goals.